src/librustc_resolve/lib.rs

```diff
@@ -2555,6 +2599,24 @@ impl<'a> Resolver<'a> {
                             function_type_rib.bindings.insert(ident, def);
                             self.record_def(param.id, PathResolution::new(def));
                         }
+                        GenericParamKind::Const { .. } => {
+                            let ident = param.ident.modern();
+                            debug!("with_type_parameter_rib: {}", param.id);
+
+                            if seen_bindings.contains_key(&ident) {
+                                let span = seen_bindings.get(&ident).unwrap();
+                                let err = ResolutionError::NameAlreadyUsedInParameterList(
+                                    ident.name,
+                                    span,
+                                );
+                                resolve_error(self, param.ident.span, err);
+                            }
+                            seen_bindings.entry(ident).or_insert(param.ident.span);
+
+                            let def = Def::ConstParam(self.definitions.local_def_id(param.id));
+                            function_type_rib.bindings.insert(ident, def);
+                            self.record_def(param.id, PathResolution::new(def));
+                        }
                     }
                 }
                 self.ribs[TypeNS].push(function_type_rib);
```

Comment:
Const parameters are defined in value namespace, not in type namespace.
(So you can also have a type parameter and const parameter with the same name.)

Reply:
Once you fix this, you'll encounter an equivalent of https://github.com/rust-lang/rust/issues/57523 though (const parameters treated as hygienic by macro_rules).
I don't suggest to fix it right now, but could you at least make an issue once this PR is merged?

Reply:
Two tests are failing now, confusing `CannotCaptureDynamicEnvironmentInFnItem` with `AttemptToUseNonConstantValueInConstant`, but I haven't been able to work out what's going wrong yet. If you can spot what the issue is, that'd be very helpful!